src/coprocessor/dag/batch/executors/stream_aggr_executor.rs

```diff
@@ -0,0 +1,515 @@
+            keys: Vec::new(),
+            states: Vec::new(),
+        };
+
+        Ok(Self(AggregationExecutor::new(
+            aggr_impl,
+            src,
+            config,
+            aggr_defs,
+            aggr_def_parser,
+        )?))
+    }
+}
+
+impl<Src: BatchExecutor> AggregationExecutorImpl<Src> for BatchStreamAggregationImpl {
+    #[inline]
+    fn prepare_entities(&mut self, entities: &mut Entities<Src>) {
+        let src_schema = entities.src.schema();
+        for group_by_exp in &self.group_by_exps {
+            entities
+                .schema
+                .push(group_by_exp.ret_field_type(src_schema).clone());
+        }
+    }
+
+    #[inline]
+    fn process_batch_input(
+        &mut self,
+        entities: &mut Entities<Src>,
+        mut input: LazyBatchColumnVec,
+    ) -> Result<()> {
+        let context = &mut entities.context;
+        let src_schema = entities.src.schema();
+
+        let rows_len = input.rows_len();
+        let group_by_len = self.group_by_exps.len();
+        let aggr_fn_len = entities.each_aggr_fn.len();
+
+        // Decode columns with mutable input first, so subsequent access to input can be immutable
+        // (and the borrow checker will be happy)
+        ensure_columns_decoded(context, &self.group_by_exps, src_schema, &mut input)?;
+        ensure_columns_decoded(context, &entities.each_aggr_exprs, src_schema, &mut input)?;
+        let group_by_results = eval_exprs(context, &self.group_by_exps, src_schema, &input)?;
+        let aggr_expr_results = eval_exprs(context, &entities.each_aggr_exprs, src_schema, &input)?;
+
+        // Stores input references, clone them when needed
+        let mut group_key = Vec::with_capacity(group_by_len);
+        let mut group_start_row = 0;
+        for row_index in 0..rows_len {
+            for group_by_result in &group_by_results {
+                // Unwrap is fine because we have verified the group by expression before.
+                let group_column = group_by_result.vector_value().unwrap();
+                group_key.push(group_column.get_unchecked(row_index));
+            }
+            match self.keys.rchunks_exact(group_by_len).next() {
+                Some(current_key) if &group_key[..] == current_key => {
+                    group_key.clear();
+                }
+                _ => {
+                    // Update the complete group
+                    if row_index > 0 {
+                        update_current_states(
+                            context,
+                            &mut self.states,
+                            aggr_fn_len,
+                            &aggr_expr_results,
+                            group_start_row,
+                            row_index,
+                        )?;
+                    }
+
+                    // create a new group
+                    group_start_row = row_index;
+                    self.keys.extend(group_key.drain(..).map(Into::into));
```

Comment:
This clone (in `Into::into`) is very implicit, since `Into` is only for Owned to Owned conversion, while here it is a reference to ownership in depth.

Reply:
I'll change it when we change the function name